### PR TITLE
Feature/metal triangle

### PR DESCRIPTION
**Pull request by LtSnail**
Add a very basic and dirty implementation of rendering a triangle

### CODE REVIEW

File: Engine/Fuego/Application.cpp [hunk 1]
```diff
@@ -26,7 +26,7 @@ Application::Application() : d(new ApplicationImpl())
     d->m_EventQueue = EventQueue::CreateEventQueue();
     d->m_Window = Window::CreateAppWindow(WindowProps(), *d->m_EventQueue);
     d->m_Running = true;
-
+
```
Comment (LepusAlleni) [inline Engine/Fuego/Application.cpp:29]: =(

File: Engine/Fuego/MacOS/Metal/Triangle.metal [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+vertex float4
+vertexShader(uint vertexID [[vertex_id]],
+             constant simd::float3* vertexPositions)
+{
+    float4 vertexOutPositions = float4(vertexPositions[vertexID][0],
+                                       vertexPositions[vertexID][1],
+                                       vertexPositions[vertexID][2],
+                                       1.0f);
+    return vertexOutPositions;
```
Comment (LepusAlleni) [inline Engine/Fuego/MacOS/Metal/Triangle.metal:13]: return float4(vertexPositions[vertexID], 1);

File: Engine/Fuego/MacOS/Metal/MetalContext.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    virtual bool Init() override;
+    virtual void SwapBuffers() override;
+    
+    // Temporary code to draw a triangle as fast as possible
+    MTL::Buffer* triangleVertexBuffer;
+    MTL::CommandQueue* metalCommandQueue;
+    MTL::RenderPipelineState* metalRenderPSO;
+    
+   private:
+    void* _window;
```
Comment (LepusAlleni) [inline Engine/Fuego/MacOS/Metal/MetalContext.h:28]: Window used to create swapchain

File: Engine/Fuego/MacOS/Metal/MetalContext.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    MTL::Device* _device;
+    CA::MetalDrawable* _metalDrawable;
```
Comment (LepusAlleni) [inline Engine/Fuego/MacOS/Metal/MetalContext.h:30]: Stored by user 
```
drawable = swapchain->acquire_image()
...
SetRenderTarget(drawable);
```

File: Engine/Fuego/MacOS/Metal/MetalContext.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    void* _metalLayer;
```
Comment (LepusAlleni) [inline Engine/Fuego/MacOS/Metal/MetalContext.h:31]: Swapchain object abstraction

File: Engine/Fuego/MacOS/Metal/MetalContext.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+namespace Fuego
+{
+class MetalContext : public GraphicsContext
+{
+   public:
+    MetalContext(void* nsWindow);
+    virtual ~MetalContext() override;
+
+    virtual bool Init() override;
+    virtual void SwapBuffers() override;
```
Comment (LepusAlleni) [inline Engine/Fuego/MacOS/Metal/MetalContext.h:20]: Split to acquire and present. Present can be done only in graphics queue. Acquire is a function of swapchain. Only one image at a time or not more than swapchain image count.

**Review by LepusAlleni — DISMISSED** [review]
Clang-format

**Review by Mon1Amour4 — APPROVED** [review]

File: Engine/Fuego/Application.cpp (hunk 1 above)
Comment (LtSnail) [reply Engine/Fuego/Application.cpp:29]: Will be addressed in the pre-commit PR

**Pull request merged**